src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -138,6 +144,58 @@ public void End(ProduceEndType endType)
             }
         }
 
+        public MemoryPoolIterator2 ProducingStart()
+        {
+            lock (_returnLock)
+            {
+                Debug.Assert(!_isProducing);
+                _isProducing = true;
+
+                if (_tail == null)
+                {
+                    throw new IOException("The socket has been closed.");
+                }
+
+                return new MemoryPoolIterator2(_tail, _tail.End);
+            }
+        }
+
+        public void ProducingComplete(MemoryPoolIterator2 end, int count)
```

Comment:
The design is to do your own appending that potentially links multiple blocks (like CopyFrom does). Once your done appending blocks, you simply pass back an iterator pointing to the end of the response data that was just appended.

Note that this design *requires* that you fill the entire previous block before linking to the `Next` block. I originally had a bunch of `Debug.Asserts` in `LockWrite` to verify this is the case, but decided this was a little noisy.